---
title: "Genetic structure"
author: "Óscar Romero Báez"
date: '2022-05-17'
output: html_document
---

This script is used to find the most probable K of the four sampling sites where the lizard *Sceloporus gramicus* was captured. First, run Admixture on the data to find out which K is most likely. Then, the cross-validation value (CV) and individual genetic assignment probability are plotted.

If don't have the Admixture and Plink softwares, can be downloaded from the following links. Admixture: https://dalexander.github.io/admixture/download.html and Plink: https://www.cog-genomics.org/plink/. 

As additional tests, the determination of the number of clusters (find.clusters) and discriminant analysis of principal components (DAPC) was performed, to see if there were agreements and additional support for Admixture analysis.

Go to the **Sgrammicus_SelectionTest** folder to work with. The first five steps must be executed in console.

#Admixture
## Create input file to run Admixture (Console)

1) Transform the .vcf database to a plink format and the output files are .ped and .map.

```{r}
#vcftools --vcf Data/Raw_Data/Sg_Pols2M_7000snps.vcf --plink --out Data/Admixture/Formats/Sgram_Plink
```

2) Insert the word "Scaffold_" at the beginning of all the lines of the .map file, which are the names of the chromosomes. Since if don't put this it will not be possible to transform it into an input file for Admixture. This using "sed" command.

```{r}
#sed -i '.bak' 's/^/Scaffold_/' Data/Admixture/Formats/Sgram_Plink.map
```

3) Transform the edited Sgram_Plink files and use the Sgram_Admixture.ped file as Admixture input.

```{r}
#plink --file Data/Admixture/Formats/Sgram_Plink --recode12 --allow-extra-chr --out Data/Admixture/Formats/Sgram_Admixture
```

4) Go to the Result folder of Admixture. Run Admixture using the following loop to get the results.

```{r}
#cd Result/Admixture/Results
#for K in 1 2 3 4 5 6 7 8 9 10; #List the number of K to test.
#do admixture --cv=10 -s 123 -B=2000 
#../../../Data/Admixture/Formats/Sgram_Admixture.ped $K -j6 | tee log${K}.out; #Set the parameters to run Admixture.
#done
```

5) Obtain the CV values (this using the grep command) and estimate the most probable K.

```{r}
#grep -h "CV" log* | sed 's/:/\ \t/' > CV_values
```

## Obtain CV plot

1) Requited libraries 

```{r}
library(vcfR)
library(adegenet)
library(reshape2)
library(gridExtra)
library(ggplot2)
library(ggpubr)
```

2) Load the necessary databases.

```{r}
#ID and location database information
Sg.data<-read.table("Data/Pop_Info/Pops.txt",sep="\t",header=T)

#SNPs database
Sg.vcf<-read.vcfR("Data/Raw_Data/Sg_Pols2M_7000snps.vcf")
#Transform SNPs database to genind object
Sg.gen<-vcfR2genind(Sg.vcf)
#Definition of localities as populations (for now)
pop(Sg.gen)<-Sg.data$loc

#CV database
(Sg.CV<-read.table("Analysis/Admixture/Results/CV_values"))

#Transform V3 from character to factor, taking each K value as level
Sg.CV$V3<-factor(x=Sg.CV$V3,levels=c("(K=1)","(K=2)","(K=3)","(K=4)","(K=5)",
                                   "(K=6)","(K=7)","(K=8)","(K=9)","(K=10)"))
```

3) Plot of CV

```{r}
(Sg.CV.plot<-ggplot(Sg.CV,aes(x=as.factor(V3),y=V4,group=1))+ #Axis definition
  geom_point(size=3)+ #Geometry definition
  geom_line()+ #Line add
  ylab("CV")+ #y-axis legend
  theme_light()+ #Canvas definition
  theme(axis.title.x=element_blank(), #Without axis title x
        axis.text.x=element_text(size=11), #Size of axis text x
        axis.title.y=element_text(size=13),#Size of axis title y
        axis.text.y=element_text(size=11))) #Axis text y
```

Save plot to pdf file

```{r}
ggsave("Analysis/Admixture/Figures/Sg.CV.plot.pdf",dpi=300,limitsize=TRUE,
       width=10,height=7)
```

The lowest values of K were, in ascending order, 3, 4, and 2, with the best value of K for my data being K = 3.

## Obtain individual genetic assignment probability plot

1) Load the Q file of each different value(s) of K with the lowest CV values.

```{r}
#K=2
Sg.K2<-read.table("Analysis/Admixture/Results/Sgram_Admixture.2.Q")

#K=3
Sg.K3<-read.table("Analysis/Admixture/Results/Sgram_Admixture.3.Q")

#K=4
Sg.K4<-read.table("Analysis/Admixture/Results/Sgram_Admixture.4.Q")
```

2) Define ID and locality information for each value of K

```{r}
#K=2
Sg.K2$ind<-Sg.data$ind
Sg.K2$loc<-Sg.data$loc

#K=3
Sg.K3$ind<-Sg.data$ind
Sg.K3$loc<-Sg.data$loc

#K=4
Sg.K4$ind<-Sg.data$ind
Sg.K4$loc<-Sg.data$loc
```

3) Generate a format compatible with ggplot

```{r}
#K=2
Sg.K2.gg<-melt(Sg.K2)

#K=3
Sg.K3.gg<-melt(Sg.K3)

#K=4
Sg.K4.gg<-melt(Sg.K4)
```

4) Plot the lowest values of K obtained of Admixture.

a) K=2
```{r}
(Sg.K2.plot<-ggplot(data=Sg.K2.gg,aes(x=ind,y=value,fill=variable))+ #Axis definition
  geom_bar(stat="identity")+ #Geometry definition
  scale_color_brewer(palette="Dark2")+ #Scale color
  facet_grid(~loc,scales="free_x",labeller=label_wrap_gen(multi_line=FALSE))+ #Panel division for location
  ylab("K2")+ #y-axis legend
  theme_minimal()+ #Canvas definition
  theme(axis.text.x=element_text(size=8,angle=90,hjust=1), #Size and position of axis.text.x
        axis.text.y=element_text(size=8),legend.position="none", #Size of axis text y and remove legend
        axis.title.x=element_blank(),axis.title.y=element_text(size=12), #Axis title
        strip.text.x=element_text(size=14))) #Size of strip text x
```

Save plot to pdf file

```{r}

ggsave("Analysis/Admixture/Figures/Sg.K2.plot.pdf",dpi=300,limitsize=TRUE,
       width=10,height=7)
```

b) K=3
```{r}
#K=3
(Sg.K3.plot<-ggplot(data=Sg.K3.gg,aes(x=ind,y=value,fill=variable))+ #Axis definition
  geom_bar(stat="identity")+ #Geometry definition
  scale_color_brewer(palette="Dark2")+ #Scale color
  facet_grid(~loc,scales="free_x",labeller=label_wrap_gen(multi_line=FALSE))+ #Panel division for location
  ylab("K3")+ #y-axis legend
  theme_minimal()+ #Canvas definition
  theme(axis.text.x=element_text(size=8,angle=90,hjust=1), #Size and position of axis.text.x
        axis.text.y=element_text(size=8),legend.position="none", #Size of axis text y and remove legend
        axis.title.x=element_blank(),axis.title.y=element_text(size=12), #Axis title
        strip.text.x=element_text(size=14))) #Size of strip text x
```

Save plot to pdf file

```{r}
ggsave("Analysis/Admixture/Figures/Sg.K3.plot.pdf",dpi=300,limitsize=TRUE,
       width=10,height=7)
```

c) K=4
```{r}
#K=4
(Sg.K4.plot<-ggplot(data=Sg.K4.gg,aes(x=ind,y=value,fill=variable))+ #Axis definition
  geom_bar(stat="identity")+ #Geometry definition
  scale_color_brewer(palette="Dark2")+ #Scale color
  facet_grid(~loc,scales="free_x",labeller=label_wrap_gen(multi_line=FALSE))+ #Panels division for location
  ylab("K4")+ #y-axis legend
  theme_minimal()+ #Canvas definition
  theme(axis.text.x=element_text(size=8,angle=90,hjust=1), #Size and position of axis.text.x
        axis.text.y=element_text(size=8),legend.position="none", #Size of axis text y and remove legend
        axis.title.x=element_blank(),axis.title.y=element_text(size=12), #Axis title
        strip.text.x=element_text(size=14))) #Size of strip text x
```

Save plot to pdf file

```{r}
ggsave("Analysis/Admixture/Figures/Sg.K4.plot.pdf",dpi=300,limitsize=TRUE,
       width=10,height=7)
```

# Determination of the number of clusters

1) Use K-means clustering to find the most likely K with a minimum BIC. The most likely K will be obtained using the find.clusters function (adegenet). To do this, create a function to add a matrix and run the function find.clusters.

```{r}
maxK<-10 #The number of K to be tested is established
myMat<-matrix(nrow=10,ncol=maxK) #Create the matrix establishing the number of rows and columns
colnames(myMat)<-1:ncol(myMat) #Give names to the columns depending on the number of each K
for(i in 1:nrow(myMat)){ #Enumerate each row
  #find.clusters parameters
  grp<-find.clusters(Sg.gen, #Data base
                     n.pca=35, #Number of axes retained in the Principal Component Analysis (PCA) step
                     choose.n.clust=FALSE, #The number of clusters shouldn't be chosen by the user
                     max.n.clust=maxK) #Maximum number of clusters to be tried, in this case is 10
  myMat[i,]<-grp$Kstat #Integrate the BIC statistics values of each K to the matrix
}
```

2) Create data frame

```{r}
my_df<-melt(myMat) #Transform matrix to an object molten data frame
colnames(my_df)[1:3]<-c("Group", "K", "BIC") #Name columns
my_df$K<-as.factor(my_df$K) #Transform K to factor
```

3) K vs BIC plot

```{r}
(Sg.K_BIC<-ggplot(my_df,aes(x=K, y=BIC))+ #Axis definition
      geom_boxplot(aes(colour = K))+ #Geometry definition
      theme_bw()+ #Canvas definition
      xlab("Number of groups (K)")+ #x-axis legend
      theme(legend.position="none")) #Hide legend box
```

Save plot to pdf file

```{r}
ggsave("Analysis/Clusters_DAPC/Figures/Sg.K_BIC.pdf",dpi=300,limitsize=TRUE,
       width=10,height=7)
```

4) Creation of database of assignment of individuals to each cluster (K)

```{r}
#Combination of mapping database with ID database.
Sg.grp<-cbind(grp$grp,Sg.data$ind)
#Name columns
colnames(Sg.grp)[1:2]<-c("K","ID")
#Transform of database to data.frame
Sg.grp<-as.data.frame(Sg.grp)
```

5) Assignment plot to each cluster (K)

```{r}
(Sg.asig<-ggplot(Sg.grp,aes(x=ID, y=K))+ #Axis definition
         geom_point(aes(colour = K))+  #Geometry definition
         scale_color_brewer(palette="Dark2")+ #Scale color
         facet_grid(~K,scales="free_x")+ #Panels division for each K within individuals
         theme_bw()+ #Canvas definition
         theme(axis.text.x=element_text(size=7.5,angle=90,hjust=0), #Size and position of axis.text.x
               axis.text.y=element_text(size=8), #Size of axis.text.x
               legend.position="none", #Remove legend
               strip.text.x=element_text(size=0), #Remove numbers of the k
               axis.title.x=element_blank(), #Remove title of axis x
               strip.background=element_blank())) #Remove rectangle
```

Save plot to pdf file

```{r}
ggsave("Analysis/Clusters_DAPC/Figures/Sg.asig.pdf",dpi=300,limitsize=TRUE,
       width=10,height=7)
```

The best K was 3 and then 4. The result is similar to Admixture.

# DAPC

1) Use the cross-validation method to find the best PCs and Da numbers

```{r,results='hide'}
pdf("Analysis/Clusters_DAPC/Figures/Sg.xval.pdf",width = 10, height=7, 
    bg = "white",colormodel = "cmyk")
(Sg.xval<-xvalDapc(tab(Sg.gen, #Data base
                  NA.method="mean"), #Change Na to mean 
                  pop(Sg.gen), #Populations
                  n.pca.max=40, #Maximum number of PCA components to retain.
                  training.set=0.9, #Proportion of data (individuals) to be used for the training set
                  result="groupMean", #Group-wise assignment success
                  center=FALSE, #Variables shouldn't be centered to mean
                  scale=FALSE, #Variables shouldn't be scaled 
                  n.rep=300, #Number of replicates to be carried out at each level of PC retention
                  n.pca=NULL, #Number of PCs determined automatically.
                  parallel="multicore", #R parallel system
                  ncpus=4L)) #Number of cores you want to use
dev.off()
```

Cross-validation DPCA report

```{r}
Sg.xval$DAPC
```

Number of PCs with best stats

```{r}
Sg.xval$`Number of PCs Achieving Highest Mean Success`
```

PC scores, lower score is better

```{r}
Sg.xval$`Root Mean Squared Error by Number of PCs of PCA`
```

n.pca = 10 and n.da = 3 is optimal. Proportion of conserved variance = 0.758.

2) DAPC using the best n.pca and n.da.

```{r}
Sg.dapc<-dapc(Sg.gen, #Data base
              pop(Sg.gen), #Populations
              n.pca=10, #Number of axes retained in the Principal Component Analysis (PCA) step
              n.da=3, #Number of axes retained in the Discriminant Analysis step
              center=FALSE, #Variables shouldn't be centered to mean
              scale=FALSE) #Variables shouldn't be scaled
```

3) DAPC plot

```{r}
#Obtain and transform to a data frame the individual coordinates of the linear discriminants of the DPCA
Sg.ind.coord<-as.data.frame(Sg.dapc$ind.coord)
#Obtain and transform the sampling sites into factors
Sg.sites<-as.factor(Sg.data$loc)
#Join both databases
Sg.scatter<-cbind(Sg.ind.coord,Sg.sites)
#Rename the columns
colnames(Sg.scatter)[1:4]<-c("LD1","LD2","LD3","Sites")
#Plot DPCA
(Sg.ggscatter<-ggscatter(Sg.scatter,x="LD1",y="LD2",col="Sites", #Axis and color definition
               ellipse=TRUE, #Add ellipse
               mean.point=F, #Remove mean point
               star.plot=TRUE)+ #Add lines in points 
               scale_color_brewer(palette="Dark2")+ #Color palette
               geom_hline(yintercept=0,linetype="dashed",color="black",size=0.5)+ #Add line in y-axis
               geom_vline(xintercept=0,linetype="dashed",color="black",size=0.5)+ #Add line in x-axis
               theme_bw()) #Canvas definition
```

Save plot to pdf file

```{r}
ggsave("Analysis/Clusters_DAPC/Figures/Sg.ggscatter.pdf",dpi=300,limitsize=TRUE,
       width=10,height=7)
```

The DAPC also represents three groups, but separates them in a slightly different way than Admixture or cluster method. Although it still shows some overlap between two sites (P5For and P5Agr), but excludes P6Rur and markedly separates P6For as well.
