---
title: "Getting input of BayPass"
author: "Óscar Romero Báez"
date: '2022-05-23'
output: html_document
---

This script is used to get the input required to run the BayPass program (Gautier M, 2015; Olazcuaga L et al., 2020). The file to start the process is the .vcf file database.

If it is required to download the program, it can be obtained at this link: http://www1.montpellier.inra.fr/CBGP/software/baypass/download.html.

Go to the **Sgrammicus_Selection** folder to work.

### 1) Required libraries

```{r,results="hide"}
library(vcfR)
library(adegenet)
```

### 2) The required databases are loaded and processed

```{r}
#SNPs database
Sg.data<-read.vcfR("Data/Loci_Data/Sg_Pols2M_7000snps.vcf")

#Transform SNPs database to genind object
Sg.gen<-vcfR2genind(Sg.data)

#ID and location database information
Sg.data<-read.table("Data/Pop_Info/Pops.txt",header=T)
#Specify your respective location
pop(Sg.gen)<-Sg.data$loc

#The database is separated by sampling site
Sg.ind.bypop<-seppop(Sg.gen)
```

### 3) Input for Baypass is generated per site

a) P5For site

```{r}
#P5For site is extracted as database
Pop_P5For<-as.data.frame(t(Sg.ind.bypop$P5For@tab))

#Separate by allele, in this case adegenet names them as "locus1.0" and "locus1.1"
Pop_P5For_Allele.0<-Pop_P5For[grep(rownames(Pop_P5For),pattern=".0$"),]
Pop_P5For_Allele.1<-Pop_P5For[grep(rownames(Pop_P5For),pattern=".1$"),]

#The alleles are put together by adding the ones that exist per population and NA is obtained
Pop_P5For.baypass<-cbind(rowSums(Pop_P5For_Allele.0,na.rm=T),rowSums(Pop_P5For_Allele.1,na.rm=T))
```

b) P5Agr site

```{r}
#P5Agr site is extracted as database
Pop_P5Agr<-as.data.frame(t(Sg.ind.bypop$P5Agr@tab))

#Separate by allele, in this case adegenet names them as "locus1.0" and "locus1.1" 
Pop_P5Agr_Allele.0<-Pop_P5Agr[grep(rownames(Pop_P5Agr),pattern=".0$"),]
Pop_P5Agr_Allele.1<-Pop_P5Agr[grep(rownames(Pop_P5Agr),pattern=".1$"),]

#The alleles are put together by adding the ones that exist per population and NA is obtained
Pop_P5Agr.baypass<-cbind(rowSums(Pop_P5Agr_Allele.0,na.rm=T),rowSums(Pop_P5Agr_Allele.1,na.rm=T))
```

c) P6For site

```{r}
#P6For site is extracted as database
Pop_P6For<-as.data.frame(t(Sg.ind.bypop$P6For@tab))

#Separate by allele, in this case adegenet names them as "locus1.0" and "locus1.1" 
Pop_P6For_Allele.0<-Pop_P6For[grep(rownames(Pop_P6For),pattern=".0$"),]
Pop_P6For_Allele.1<-Pop_P6For[grep(rownames(Pop_P6For),pattern=".1$"),]

#The alleles are put together by adding the ones that exist per population and NA is obtained
Pop_P6For.baypass<-cbind(rowSums(Pop_P6For_Allele.0,na.rm=T),rowSums(Pop_P6For_Allele.1,na.rm=T))
```

d) P6Agr site

```{r}
#P6Agr site is extracted as database
Pop_P6Agr<-as.data.frame(t(Sg.ind.bypop$P6Agr@tab))

#Separate by allele, in this case adegenet names them as "locus1.0" and "locus1.1" 
Pop_P6Agr_Allele.0<-Pop_P6Agr[grep(rownames(Pop_P6Agr),pattern=".0$"),]
Pop_P6AgrAllele.1<-Pop_P6Agr[grep(rownames(Pop_P6Agr),pattern=".1$"),]

#The alleles are put together by adding the ones that exist per population and NA is obtained
Pop_P6Agr.baypass<-cbind(rowSums(Pop_P6Agr_Allele.0,na.rm=T),rowSums(Pop_P6AgrAllele.1,na.rm=T))
```

### 4) All bases are combined using the `cbind()` function

a) All sites

```{r}
#Combination of databases from all sites
SN_formato_all.baypass<-cbind(Pop_P5For.baypass,Pop_P5Agr.baypass,Pop_P6For.baypass,Pop_P6Agr.baypass)

#Input for BayPass is written and saved
write.table(SN_formato_all.baypass,file="Data/BayPass/Input_Loci/SN_baypass_all.input",row.names=F,col.names=F,sep=" ",quote=FALSE)

#It is useful to have the numbered list of the loci, to know which is which when you have results. The format is written and saved
write.table(rownames(SN_formato_all.baypass),file="Data/BayPass/Data_Loci/SN_baypass_all.loci",row.names=T,col.names=F,sep=" ",quote=FALSE)
```

b) Only P6 sites

```{r}
#Combination of databases from P6 sites
SN_formato_P6.baypass<-cbind(Pop_P6For.baypass,Pop_P6Agr.baypass)

#Input for BayPass is written and saved
write.table(SN_formato_P6.baypass,file="Data/BayPass/Input_Loci/SN_baypass_P6.input",row.names=F,col.names=F,sep=" ",quote=FALSE)

#It is useful to have the numbered list of the loci, to know which is which when you have results. The format is written and saved
write.table(rownames(SN_formato_P6.baypass),file="Data/BayPass/Data_Loci/SN_baypass_P6.loci",row.names=T,col.names=F,sep=" ",quote=FALSE)
```

c) Only agricultural sites

```{r}
#Combination of databases from P5 sites
SN_formato_P5.baypass<-cbind(Pop_P5For.baypass,Pop_P5Agr.baypass)

#Input for BayPass is written and saved
write.table(SN_formato_P5.baypass,file="Data/BayPass/Input_Loci/SN_baypass_P5.input",row.names=F,col.names=F,sep=" ",quote=FALSE)

#It is useful to have the numbered list of the loci, to know which is which when you have results. The format is written and saved
write.table(rownames(SN_formato_P5.baypass),file="Data/BayPass/Data_Loci/SN_baypass_P5.loci",row.names=T,col.names=F,sep=" ",quote=FALSE)
```
